sysl2/sysl/deps.go

```diff
@@ -0,0 +1,264 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	"github.com/yale8848/stream"
+)
+
+type AppElement struct {
+	Name     string
+	Endpoint string
+}
+
+type AppDependency struct {
+	Self      *AppElement
+	Target    *AppElement
+	Statement *sysl.Statement
+}
+
+type DependencySet struct {
+	Deps map[string]*AppDependency
+}
+
+func (dep *AppDependency) String() string {
+	return fmt.Sprintf("%s:%s:%s:%s", dep.Self.Name, dep.Self.Endpoint, dep.Target.Name, dep.Target.Endpoint)
+}
+
+func (ds *DependencySet) ToSlice() []*AppDependency {
+	o := []*AppDependency{}
+	keys := []string{}
+	for k := range ds.Deps {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		o = append(o, ds.Deps[k])
+	}
+
+	return o
+}
+
+func (dep *AppDependency) extractAppNames() StrSet {
+	s := StrSet{}
+	s.Insert(dep.Self.Name)
+	s.Insert(dep.Target.Name)
+	return s
+}
+
+func (dep *AppDependency) extractEndpoints() StrSet {
+	s := StrSet{}
+	s.Insert(dep.Self.Endpoint)
+	s.Insert(dep.Target.Endpoint)
+	return s
+}
+
+func NewDependencySet() *DependencySet {
+	return &DependencySet{map[string]*AppDependency{}}
+}
+
+func MakeAppDependency(self, target *AppElement, stmt *sysl.Statement) *AppDependency {
+	return &AppDependency{self, target, stmt}
+}
+
+func MakeAppElement(name, endpoint string) *AppElement {
+	return &AppElement{name, endpoint}
+}
+
+type CallStatementSlice struct {
+	slice []*sysl.Statement
+}
+
+func NewCallStatementSlice() *CallStatementSlice {
+	return &CallStatementSlice{}
+}
+
+func (cs *CallStatementSlice) Add(call *sysl.Statement) {
+	cs.slice = append(cs.slice, call)
+}
+
+func (cs *CallStatementSlice) GetSlice() []*sysl.Statement {
+	return cs.slice
+}
+
+func (cs *CallStatementSlice) CollectCallStatements(stmts []*sysl.Statement) {
+	for _, stat := range stmts {
+		switch c := stat.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			cs.Add(stat)
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			cs.CollectCallStatements(c.Cond.GetStmt())
+		case *sysl.Statement_Loop:
+			cs.CollectCallStatements(c.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			cs.CollectCallStatements(c.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			cs.CollectCallStatements(c.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			cs.CollectCallStatements(c.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range c.Alt.GetChoice() {
+				cs.CollectCallStatements(choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func FindApps(module *sysl.Module, excludes, integrations StrSet, ds []*AppDependency, highlight bool) StrSet {
+	r := []string{}
+	appReStr := toPattern(integrations.ToSlice())
+	re := regexp.MustCompile(appReStr)
+	st, _ := stream.Of(ds)
+	st.ForEach(func(v stream.T) bool {
+		if dep, ok := v.(*AppDependency); ok {
```

Comment:
@JackLeeHal I dont see any reason why this cannot be done using `for range slice`.
Please remove this `stream` dependency.

Reply:
provide better names for variables
what is `r`?

Reply:
removed.

Reply:
changed r to output.